whats-that-typeclass-alternative/article.md

```diff
@@ -0,0 +1,414 @@
+First of all, because of the possibility to compose multiple `Parser i a` with
+the `<|>` operator.
+
+Another reason is two methods of `Alternative` that we haven't
+discussed so far, which we get for free:
+
+* `some :: f a -> f [a]` which means: _one or more_
+* `many :: f a -> f [a]` which means: _zero of more_
+
+In our case, `some (parseLT 10)` will create a new parser of type
+`parseLT :: Int -> Parser Int [Int]` that expects at least one integer
+that is smaller 10 and then will consume all following numbers which are
+smaller 10 as well.
+
+In case of `many (parseLT 10)`, the type signature looks the same as in the
+`some` case, but it does not fail if the first element in the input isn't a
+number smaller 10. In that case, it will just return the empty list `[]`.
+
+So lets try out these methods:
+
+```none
+λ> evalParser (some $ parseLT 10) [2,3,10,20]
+Right [2,3]
+
+λ> evalParser (many $ parseLT 10) [2,3,10,20]
+Right [2,3]
+
+λ> evalParser (some $ parseLT 10) [10,2,3,20]
+Left "parser failed on [10,2,3,20]"
+
+λ> evalParser (many $ parseLT 10) [10,2,3,20]
+Right []
+```
+
+`some` and `many` are a very common and useful pattern in parser combinators.
+
+If you want to dive deeper into the topic of parsers in Haskell, take a look
+[at this post](https://serokell.io/blog/parser-combinators-in-haskell).
+
+## Conclusion and outlook
+
+* `Alternative` is a useful instance to implement for your applicative functor
+  if it has a semantic of _try this or alternatively that_.
+* `Alternative` can be seen as a monoid on applicative functors.
+* If you implement a parser (combinator), you almost always want to implement
+  an `Alternative` instance for it.
+
+Another very closely related typeclass is [`MonadPlus m`](https://wiki.haskell.org/MonadPlus)
+which has the same semantics as `Alternative` but for types which also implement `Monad`. 
+So it can be seen as a "monoid on monads". It comes with a default implementation
+based on `Alternative`, so you don't have to actually implement any methods:
+
+```haskell
+-- The MonadPlus class definition
+
+-- | Monads that also support choice and failure.
+class (Alternative m, Monad m) => MonadPlus m where
+   -- | The identity of 'mplus'.  It should also satisfy the equations
+   --
+   -- > mzero >>= f  =  mzero
+   -- > v >> mzero   =  mzero
+   --
+   -- The default definition is
+   --
+   -- @
+   -- mzero = 'empty'
+   -- @
+   mzero :: m a
+   mzero = empty
+
+   -- | An associative operation. The default definition is
+   --
+   -- @
+   -- mplus = ('<|>')
+   -- @
+   mplus :: m a -> m a -> m a
+   mplus = (<|>)
+```
+
+## Exercises
+
+1. Implement the `isAliceOrBobOrCharly` example from the beginning in terms of
+   our new `Parser i a` type.
+
+   So you come up with a parser:
+
+   ```haskell
+   parseAliceBobCarly :: Parser Char String
+   parseAliceBobCarly = ...
+   ```
+
+   You are allowed to utilize functions from `Data.List`
```

Comment:
Do you think it's worth it to talk about `Data.Foldable.asum`? It's a useful thing that you frequently run across.

Reply:
Yes it is, linked it and provided an alternative implemenation using `asum`